compiler/rustc_ast_lowering/src/pat.rs

```diff
@@ -429,4 +429,80 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         };
         self.arena.alloc(hir::PatExpr { hir_id: self.lower_node_id(expr.id), span, kind })
     }
+
+    pub(crate) fn lower_ty_pat(&mut self, pattern: &Pat) -> &'hir hir::TyPat<'hir> {
+        self.arena.alloc(self.lower_ty_pat_mut(pattern))
+    }
+
+    fn lower_ty_pat_mut(&mut self, mut pattern: &Pat) -> hir::TyPat<'hir> {
+        // loop here to avoid recursion
+        let pat_hir_id = self.lower_node_id(pattern.id);
+        let node = loop {
+            match &pattern.kind {
+                PatKind::Range(e1, e2, Spanned { node: end, .. }) => {
+                    let mut lower_expr = |e: &Expr| -> &_ {
+                        let kind = if let ExprKind::Path(qself, path) = &e.kind {
+                            hir::ConstArgKind::Path(self.lower_qpath(
+                                e.id,
+                                qself,
+                                path,
+                                ParamMode::Optional,
+                                AllowReturnTypeNotation::No,
+                                ImplTraitContext::Disallowed(ImplTraitPosition::Path),
+                                None,
+                            ))
+                        } else {
+                            let node_id = self.next_node_id();
+                            let def_id = self.create_def(
+                                self.current_hir_id_owner.def_id,
+                                node_id,
+                                kw::Empty,
+                                DefKind::AnonConst,
+                                e.span,
+                            );
+                            let hir_id = self.lower_node_id(node_id);
+                            let ac = self.arena.alloc(hir::AnonConst {
+                                def_id,
+                                hir_id,
+                                body: self.lower_const_body(pattern.span, Some(e)),
+                                span: self.lower_span(pattern.span),
+                            });
+                            hir::ConstArgKind::Anon(ac)
+                        };
```

Comment:
this is pretty much mgce? :thinking: as in, this also adds consts in the type system.

Can you delegate this to lower_const_arg or whatever instead?

Reply:
Yes, but at that point I should modify the ast, and I'd prefer to do that in other PRs

Reply:
please add a FIXME then :grin: